### PR TITLE
Task/05 hooks

### PR DESCRIPTION
**Pull request by demasha0908**
https://demasha0908.github.io/fe-react-2024/

Acceptance Criteria for 1
 

- [x] Mentor is added to pr
- [x]  AC copied to pr description
- [x]  Link to deployed version is added to pr
- [x]  Simple Pagination component (without ...) is added and works correctly
- [x]  On the 1st page button Back should be disabled, same as button Forward on the last page
- [x]  Interaction with pagination should not change or break sorting and filtration.
- [x]  Category filter (multi-select) is added and works correctly
- [x]  Sorting input (single select) is added and works correctly
- [x]  Interaction with filter and sorter should reset pagination and move a user to the 1st page

Acceptance Criteria for 1*

- [x]  Pagination component with ... is implemented according to [design](
- [x]  Search input works correctly

### CODE REVIEW

**Comment by demasha0908** [comment]
Sorry, I don't know how to fix error :(

File: src/components/searchbar/SearchBar.component.tsx [hunk 1]
```diff
@@ -1,10 +1,130 @@
...
+                    break;
+                }
+                case 'oldest': {
+                    filtered = filtered.sort((a, b) => new Date(a.creationAt).getTime() - new Date(b.creationAt).getTime());
+                    break;
+                }
+            }
+        }
+        onFilteredProducts(filtered);
+    }, [activeFilter, searchTerm, activeCategories, products, currentPage]);
```
Comment (vladyslav-yermolin-moc) [inline src/components/searchbar/SearchBar.component.tsx:58]: quick fix:
```suggestion
    }, [activeFilter, searchTerm, activeCategories, products, currentPage, onFilteredProducts]);
```
Linter said: `onFilteredProducts` is a variable (function in this case) you are using inside the `useEffect` hook. So you should add this variable to the dependencies of the hook.

**Comment by vladyslav-yermolin-moc** [comment]
Thanks for your pr, will check it during the day

File: src/components/product/Pagination.module.css [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  box-shadow: 0 0 0 1px var(--pagehover);
+  color: var(--text-maincolor);
+  border-radius: 8px;
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  padding: 0;
+}
+
+.arrowButton:hover {
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.module.css:21]: Please correct the color of disabled `>` - should be `#ccc`
Also would be good to block any interaction with disabled button - not change bg-color on hover and change the cursor to default
image

File: src/components/product/Pagination.component.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            onPageChange(newPage);
+        }
+    };
+
+    const handleNextClick = () => {
+        if (activePage < totalPages) {
+            const newPage = activePage + 1;
+            setActivePage(newPage);
+            onPageChange(newPage);
+        }
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.component.tsx:68]: Using the hooks inside of the conditions is a bad practice.
```suggestion
        if (activePage >= totalPages) {
          return;
        }
            const newPage = activePage + 1;
            setActivePage(newPage);
            onPageChange(newPage);
```

File: src/components/product/Pagination.component.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        setActivePage(page as number);
+        onPageChange(page as number);
+    };
+
+    const handlePreviousClick = () => {
+        if (activePage > 1) {
+            const newPage = activePage - 1;
+            setActivePage(newPage);
+            onPageChange(newPage);
+        }
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.component.tsx:60]: same here

File: src/components/product/Pagination.component.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            <button onClick={handlePreviousClick} className={styles.arrowButton} disabled={activePage === 1}>
+                <LeftArrowIcon disable={activePage === 1} />
+            </button>
+
+            {getPagination().map((page, index) => (
+                <button
+                    key={index}
+                    onClick={() => handleClick(page)}
+                    disabled={typeof page === 'string'}
+                    className={
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.component.tsx:82]: would be good to move this logic to separate method or use [classNames](https://github.com/JedWatson/classnames)

File: src/components/product/Pagination.component.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+interface PaginationProps {
+    totalPages: number;
+    currentPage: number;
+    onPageChange: (page: number) => void;
+}
+
+export const Pagination: React.FC<PaginationProps> = ({ totalPages, onPageChange, currentPage }) => {
+    const [activePage, setActivePage] = useState<number>(currentPage);
+
+    const getPagination = (): (string | number)[] => {
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.component.tsx:17]: Would be good to move outside of the component, into some `helper.ts`

File: src/components/product/Pagination.component.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        <div className={styles.paginationWrapper}>
+            <button onClick={handlePreviousClick} className={styles.arrowButton} disabled={activePage === 1}>
+                <LeftArrowIcon disable={activePage === 1} />
+            </button>
+
+            {getPagination().map((page, index) => (
+                <button
+                    key={index}
+                    onClick={() => handleClick(page)}
+                    disabled={typeof page === 'string'}
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.component.tsx:81]: This condition is implicit. Make it explicit
```suggestion
const DOTS = '...'

                    disabled={page === DOTS}
```

File: src/components/product/Pagination.component.tsx
Comment (vladyslav-yermolin-moc) [inline src/components/product/Pagination.component.tsx:1]: Looks like you need separate folder `pagination` for these two files

File: src/components/product/ProductCard.component.tsx [hunk 8]
```diff
@@ -1,3 +1,5 @@
+import React from 'react';
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/ProductCard.component.tsx:1]: Do you need this here?

File: src/components/product/ProductList.component.tsx [hunk 9]
```diff
@@ -1,24 +1,55 @@
...
+        fetch(apiUrl)
+            .then((response) => response.json())
+            .then((data) => {
+                setProducts(data);
+                const paginated = getPaginatedProducts(data);
+                setPaginatedProducts(paginated);
+
+                setTotalPages(Math.ceil(data.length / 8));
+            });
+    });
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/ProductList.component.tsx:28]: Please add a dependecies - now useEffect makes a request on every rerender and DDOSes our backend)))
image

File: src/components/product/Products.component.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/Products.component.tsx:1]: You can skip this line everywhere https://stackoverflow.com/questions/71725865/do-i-need-to-import-react-in-every-file

**Review by vladyslav-yermolin-moc — CHANGES_REQUESTED** [review]
So good so far 👍

File: src/components/product/ProductList.component.tsx [hunk 11]
```diff
@@ -1,24 +1,58 @@
...
+    }, [products, currentPage]);
+
+    const handlePageChange = (page: number) => {
+        setCurrentPage(page);
+    };
+
+    const handleFilteredProducts = (filteredProducts: Product[]) => {
+        setPaginatedProducts(filteredProducts);
+        const paginated = getPaginatedProducts(filteredProducts, currentPage);
+        setPaginatedProducts(paginated);
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/ProductList.component.tsx:42]: you set state two times

File: src/components/product/ProductList.component.tsx [hunk 12]
```diff
@@ -1,24 +1,58 @@
...
-    </section>
-);
-
-export default ProductList;
+export default function ProductList() {
+    const apiUrl = 'https://ma-backend-api.mocintra.com/api/v1/products';
+    const [products, setProducts] = useState<Product[]>([]);
+    const [totalPages, setTotalPages] = useState(0);
+
+    const [paginatedProducts, setPaginatedProducts] = useState<Product[]>([]);
```
Comment (vladyslav-yermolin-moc) [inline src/components/product/ProductList.component.tsx:17]:
```suggestion
    const [shownProducts, ...
```
or
```suggestion
    const [productsOnPage, ...```

File: src/components/searchbar/SearchBar.component.tsx
Comment (vladyslav-yermolin-moc) [inline src/components/searchbar/SearchBar.component.tsx:1]: in this component not only SearchBar. Proly Filters is a right name

Comment (vladyslav-yermolin-moc) [reply src/components/searchbar/SearchBar.component.tsx:1]: Lift up all logic from here.
Idealy would be great to split this component into 3 pieces:
- search
- sort
- category
Every component should get state and aetter from parent, like `selectedCategories` and `setSelectedCategories`. And only show selected cats and send to the parent event of selecting

**Review by vladyslav-yermolin-moc — CHANGES_REQUESTED** [review]

File: src/components/searchbar/SearchBar.component.tsx
Comment (demasha0908) [reply src/components/searchbar/SearchBar.component.tsx:1]: fixed

**Review by vladyslav-yermolin-moc — APPROVED** [review]

**Pull request merged**